### PR TITLE
Add search pipeline documentation

### LINKED ISSUE #2560: [DOC] Search Processor (new feature)
**What do you want to do?**
 
- [ ] Request a change to existing documentation
- [x] Add new documentation
- [ ] Report a technical problem with the documentation
- [ ] Other 

**Tell us about your request.** Provide a summary of the request and all versions that are affected.
 Search processor is a chain of processors to take in queries preprocess the query, do information retrieval, and rerank/enhance retrieved results.

**What other resources are available?** Provide links to related issues, POCs, steps for testing, etc.
markcoh is the SDM, SDE is froh@ and Stavros is the PM

This was targeted for 2.6, but may go experimental in 2.6 and GA in 2.7

Associated issue: https://github.com/opensearch-project/search-processor/issues/80

### LINKED ISSUE #4155: [DOC] Search pipeline script request processor
**What do you want to do?**
 
- [ ] Request a change to existing documentation
- [x] Add new documentation
- [ ] Report a technical problem with the documentation
- [ ] Other 

**Tell us about your request.** Provide a summary of the request and all versions that are affected.

A subtask of [search pipeline](https://github.com/opensearch-project/documentation-website/issues/2560) is to add docs for the script processor, which allows users to manipulate search request via painless script. We would like to explicitly mention all the [supported search request body fields ](https://github.com/opensearch-project/OpenSearch/pull/7607#issue-1714286320) in the public document. 

**What other resources are available?** Provide links to related issues, POCs, steps for testing, etc.

Meta issue https://github.com/opensearch-project/documentation-website/issues/2560

### PR DESCRIPTION
**Pull request by kolchfa-aws**
Fixes #2560 
Fixes #4155 

### Checklist
- [ x] By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license and subject to the [Developers Certificate of Origin](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: _search-plugins/search-pipelines/index.md [hunk 1]
```diff
@@ -0,0 +1,312 @@
...
+PUT /my_index/_settings 
+{
+  "index.search.default_pipeline" : "_none"
+}
+```
+{% include copy-curl.html %}
+
+## Updating a search pipeline
+
+You can update a search pipeline dynamically using the Search Pipeline API. To update an existing search processor, make sure to specify its ID in the `tag` field.
```
[the remark this answers is not in the document]
Comment (kolchfa-aws) [reply _search-plugins/search-pipelines/index.md:276]: Reworded. Having multiple processors of the same type is mentioned in the "create pipeline" section (line 45).

**Comment by kolchfa-aws** [comment]
@msfroh I've addressed the feedback. Could you please review again and approve if everything looks good? Thanks!

File: _search-plugins/search-pipelines/index.md [hunk 2]
```diff
@@ -0,0 +1,387 @@
...
+    }
+  ]
+}
+```
+{% include copy-curl.html %}
+
+In subsequent search responses, the version is provided:
+
+```json
+GET _search/pipeline/my_pipeline
```
Comment (msfroh) [inline _search-plugins/search-pipelines/index.md:358]: Not really a "search response" here. Maybe "get pipeline response" ? (as in, "the response to a get pipeline request").

**Review by msfroh — APPROVED** [review]

File: _search-plugins/search-pipelines/index.md [hunk 3]
```diff
@@ -0,0 +1,387 @@
...
+layout: default
+title: Search pipelines
+nav_order: 100
+has_children: true
+has_toc: false
+---
+
+# Search pipelines
+
+Search pipelines is an experimental feature. Therefore, we do not recommend the use of search pipelines in a production environment. For updates on the progress of search pipelines, or if you want to leave feedback that could help improve the feature, see the associated [GitHub issue](https://github.com/opensearch-project/OpenSearch/issues/6278).
```
Comment (vagimeli) [inline _search-plugins/search-pipelines/index.md:11]:
```suggestion
Search pipelines is an experimental feature and therefore is not recommended for use in a production environment. For updates on the progress of search pipelines, or if you want to leave feedback that could help improve the feature, see the associated [GitHub issue](https://github.com/opensearch-project/OpenSearch/issues/6278).    
```

File: _search-plugins/search-pipelines/index.md [hunk 4]
```diff
@@ -0,0 +1,387 @@
...
+{: .warning}
+
+To integrate result rerankers, query rewriters, and other components that operate on queries or results, you can use _search pipelines_. Search pipelines make it easier for you to process search queries and search results right in OpenSearch, as opposed to building plugins or creating additional logic elsewhere in your application. Thus, you can customize search results without adding complexity to your application. Search pipelines also provide a mechanism for modularizing your processors so you can add and reorder them easily. 
+
+## Terminology
+
+The following is a list of search pipeline terminology:
+
+* _Search request processor_: a component that takes a search request (the query and the metadata passed in the request), performs some operation with or on the search request, and returns a search request.
+* _Search response processor_: a component that takes a search response and search request (the query, results, and metadata passed in the request), performs some operation with or on the search response, and returns a search response.
```
Comment (vagimeli) [inline _search-plugins/search-pipelines/index.md:21]:
```suggestion
* _Search response processor_: A component that takes a search response and search request (the query, results, and metadata passed in the request), performs some operation with or on the search response, and returns a search response.
```

File: _search-plugins/search-pipelines/index.md [hunk 5]
```diff
@@ -0,0 +1,387 @@
...
+* _Processor_: either a search request processor or a search response processor.
```
Comment (vagimeli) [inline _search-plugins/search-pipelines/index.md:22]:
```suggestion
* _Processor_: Either a search request processor or a search response processor.
```

File: _search-plugins/search-pipelines/index.md [hunk 6]
```diff
@@ -0,0 +1,387 @@
...
+* _Search pipeline_: an ordered list of processors that is integrated into OpenSearch. The pipeline intercepts a query, performs processing on the query, sends it to OpenSearch, intercepts the results, performs processing on the results, and returns them to the calling application, as shown in the following diagram.
```
Comment (vagimeli) [inline _search-plugins/search-pipelines/index.md:23]:
```suggestion
* _Search pipeline_: An ordered list of processors that is integrated into OpenSearch. The pipeline intercepts a query, performs processing on the query, sends it to OpenSearch, intercepts the results, performs processing on the results, and returns them to the calling application, as shown in the following diagram. 
```

**Review by vagimeli — APPROVED** [review]

File: _search-plugins/search-pipelines/index.md (hunk 3 above)
Comment (kolchfa-aws) [reply _search-plugins/search-pipelines/index.md:11]: I've used this wording before but let's ask @natebower's opinion if it should be changed.

File: _search-plugins/search-pipelines/script-processor.md [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+- `terminate_after` 
+- `profile` 
+
+For request field definitions, see [search request fields]({{site.url}}{{site.baseurl}}/api-reference/search#request-body).
+
+## Request fields
+
+The following table lists all available request fields.
+
+`source` | Inline script | The script to run. Required.
```
[the remark this answers is not in the document]
Comment (kolchfa-aws) [reply _search-plugins/search-pipelines/script-processor.md:34]: Yes! Thank you

File: _search-plugins/search-pipelines/index.md [hunk 8]
```diff
@@ -0,0 +1,387 @@
...
+{: .warning}
+
+To integrate result rerankers, query rewriters, and other components that operate on queries or results, you can use _search pipelines_. Search pipelines make it easier for you to process search queries and search results right in OpenSearch, as opposed to building plugins or creating additional logic elsewhere in your application. Thus, you can customize search results without adding complexity to your application. Search pipelines also provide a mechanism for modularizing your processors so you can add and reorder them easily. 
+
+## Terminology
+
+The following is a list of search pipeline terminology:
+
+* _Search request processor_: A component that takes a search request (the query and the metadata passed in the request), performs some operation with or on the search request, and returns a search request.
+* _Search response processor_: A component that takes a search response and search request (the query, results, and metadata passed in the request), performs some operation with or on the search response, and returns a search response.
```
Comment (natebower) [inline _search-plugins/search-pipelines/index.md:21]:
```suggestion
* _Search response processor_: A component that takes a search response and search request (the query, results, and metadata passed in the request), performs an operation with or on the search response, and returns a search response.
```

File: _search-plugins/search-pipelines/index.md [hunk 9]
```diff
@@ -0,0 +1,387 @@
...
+* _Processor_: Either a search request processor or a search response processor.
+* _Search pipeline_: An ordered list of processors that is integrated into OpenSearch. The pipeline intercepts a query, performs processing on the query, sends it to OpenSearch, intercepts the results, performs processing on the results, and returns them to the calling application, as shown in the following diagram. 
+
+![Search processor diagram]({{site.url}}{{site.baseurl}}/images/search-pipelines.png)
+
+Both request and response processing for the pipeline are performed on the coordinator node so there is no shard-level processing.
```
Comment (natebower) [inline _search-plugins/search-pipelines/index.md:27]:
```suggestion
Both request and response processing for the pipeline are performed on the coordinator node, so there is no shard-level processing.
```

Comment (natebower) [inline _search-plugins/search-pipelines/index.md:27]: Tip: If the meaning of "so" is "so that," it doesn't take a comma, but if it means "therefore," as it does here, it takes a comma.

File: _search-plugins/search-pipelines/index.md [hunk 10]
```diff
@@ -0,0 +1,387 @@
...
+## Viewing available processor types
+
+You can use the Nodes Search Pipelines API to view the available processor types:
+
+```json
+GET /_nodes/search_pipelines
+```
+{% include copy-curl.html %}
+
+The response contains the `search_pipelines` object that lists the possible request and response processors:
```
Comment (natebower) [inline _search-plugins/search-pipelines/index.md:52]: "available" instead of "possible"?

File: _search-plugins/search-pipelines/index.md [hunk 11]
```diff
@@ -0,0 +1,387 @@
...
+}
+```
+</details>
+
+In addition to the processors provided by OpenSearch, additional processors may be provided by plugins.
+{: .note}
+
+## Creating a search pipeline
+
+Search pipelines are stored in the cluster state. To create a search pipeline, you must configure an ordered list of processors in your OpenSearch cluster. You can have more than one processor of the same type in the pipeline. Each processor has a `tag` identifier that distinguishes it from the others. Tagging a specific processor can be helpful for error messages, especially if you add multiple processors of the same type.
```
Comment (natebower) [inline _search-plugins/search-pipelines/index.md:113]: The meaning of "for" in "for error messages" is a little ambiguous. Should it just be "in", or is there additional clarification we can add (in what way is it helpful)?

File: _search-plugins/search-pipelines/index.md [hunk 12]
```diff
@@ -0,0 +1,387 @@
...
+      }
+    }
+  ]
+}
+```
+{% include copy-curl.html %}
+
+## Retrieving search pipelines
+
+To retrieve existing search pipeline details, use the Search Pipeline API.
```
Comment (natebower) [inline _search-plugins/search-pipelines/index.md:141]: Do we mean "To retrieve the details of an existing search pipeline"? As written, the details are doing the existing.

File: _search-plugins/search-pipelines/index.md [hunk 13]
```diff
@@ -0,0 +1,387 @@
...
+```json
+GET /my_index/_search?search_pipeline=my_pipeline
+```
+{% include copy-curl.html %}
+
+For a complete example of using a search pipeline with a `filter_query` processor, see [`filter_query` processor example]({{site.url}}{{site.baseurl}}/search-plugins/search-pipelines/filter-query-processor#example).
+
+## Default search pipeline
+
+For convenience, you can set a default search pipeline on an index. Once your index has a default pipeline, you don't need to specify the `search_pipeline` query parameter with every search request.
```
Comment (natebower) [inline _search-plugins/search-pipelines/index.md:206]:
```suggestion
For convenience, you can set a default search pipeline for an index. Once your index has a default pipeline, you don't need to specify the `search_pipeline` query parameter in every search request.
```

File: _search-plugins/search-pipelines/index.md [hunk 14]
```diff
@@ -0,0 +1,387 @@
...
+
+### Setting a default search pipeline on an index
```
Comment (natebower) [inline _search-plugins/search-pipelines/index.md:208]:
```suggestion
### Setting a default search pipeline for an index
```

File: _search-plugins/search-pipelines/index.md [hunk 15]
```diff
@@ -0,0 +1,387 @@
...
+
+To set a default search pipeline on an index, specify the `index.search.default_pipeline` in the index's settings:
```
Comment (natebower) [inline _search-plugins/search-pipelines/index.md:210]:
```suggestion
To set a default search pipeline for an index, specify the `index.search.default_pipeline` in the index's settings:
```

File: _search-plugins/search-pipelines/index.md [hunk 16]
```diff
@@ -0,0 +1,387 @@
...
+
+```json
+PUT /my_index/_settings 
+{
+  "index.search.default_pipeline" : "my_pipeline"
+}
+```
+{% include copy-curl.html %}
+
+After setting the default pipeline on `my_index`, you can try the same search for all documents:
```
Comment (natebower) [inline _search-plugins/search-pipelines/index.md:220]:
```suggestion
After setting the default pipeline for `my_index`, you can try the same search for all documents:
```

File: _search-plugins/search-pipelines/filter-query-processor.md [hunk 17]
```diff
@@ -0,0 +1,178 @@
...
+The `filter_query` search request processor intercepts a search request and applies an additional query to the request, filtering the results. This is useful when you don't want to rewrite existing queries in your application but need additional filtering of the results.
+
+## Request fields
+
+The following table lists all available request fields.
+
+Field | Data type | Description
+:--- | :--- | :---
+`query` | Object | A query in query domain-specific language. For a list of OpenSearch query types, see [Query DSL]({{site.url}}{{site.baseurl}}/opensearch/query-dsl/). Required. 
+`tag` | String | The processor's identifier. Optional.
```
Comment (macohen) [inline _search-plugins/search-pipelines/filter-query-processor.md:24]: should we say why this is useful?

File: _search-plugins/search-pipelines/filter-query-processor.md [hunk 18]
```diff
@@ -0,0 +1,178 @@
...
+{
+  "message": "This is a private message", 
+  "visibility": "private"
+}
+```
+{% include copy-curl.html %}
+
+### Creating a search pipeline 
+
+The following request creates a search pipeline with a `filter_query` request processor that uses a term query to return only public messages:
```
Comment (macohen) [inline _search-plugins/search-pipelines/filter-query-processor.md:55]: should this be '...creates a search pipeline called "my_pipeline" with a...'

File: _search-plugins/search-pipelines/index.md (hunk 3 above)
Comment (macohen) [reply _search-plugins/search-pipelines/index.md:11]: We reopened the RFC for this feature: https://forum.opensearch.org/t/rfc-search-pipelines/12099 to use instead of the GH issue.

File: _search-plugins/search-pipelines/index.md [hunk 19]
```diff
@@ -0,0 +1,387 @@
...
+      }
+    }
+  ]
+}
+```
+{% include copy-curl.html %}
+
+## Search pipeline versions
+
+You can specify a version for your pipeline when creating it in the `version` parameter:
```
Comment (natebower) [inline _search-plugins/search-pipelines/index.md:334]: Do we mean "When creating your pipeline, you can specify a version for it in the `version` parameter"?

File: _search-plugins/search-pipelines/rename-field-processor.md [hunk 20]
```diff
@@ -0,0 +1,214 @@
...
+title: Rename field processor
+nav_order: 20
+has_children: false
+parent: Search pipelines
+grand_parent: Search
+---
+
+# Rename field processor
+
+Rename field processor is part of search pipeline functionality. Search pipelines is an experimental feature. For updates on the progress of search pipelines, or if you want to leave feedback that could help improve the feature, see the associated [GitHub issue](https://github.com/opensearch-project/OpenSearch/issues/6278).
```
Comment (natebower) [inline _search-plugins/search-pipelines/rename-field-processor.md:12]:
```suggestion
The rename field processor is part of the functionality of a search pipeline, which is an experimental feature. For updates on the progress of the feature or if you want to leave feedback, see the associated [GitHub issue](https://github.com/opensearch-project/OpenSearch/issues/6278).    
```

File: _search-plugins/search-pipelines/index.md [hunk 21]
```diff
@@ -0,0 +1,387 @@
...
+has_children: true
+has_toc: false
+---
+
+# Search pipelines
+
+Search pipelines is an experimental feature. Therefore, we do not recommend the use of search pipelines in a production environment. For updates on the progress of search pipelines, or if you want to leave feedback that could help improve the feature, see the associated [GitHub issue](https://github.com/opensearch-project/OpenSearch/issues/6278).    
+{: .warning}
+
+To integrate result rerankers, query rewriters, and other components that operate on queries or results, you can use _search pipelines_. Search pipelines make it easier for you to process search queries and search results right in OpenSearch, as opposed to building plugins or creating additional logic elsewhere in your application. Thus, you can customize search results without adding complexity to your application. Search pipelines also provide a mechanism for modularizing your processors so you can add and reorder them easily.
```
Comment (macohen) [inline _search-plugins/search-pipelines/index.md:14]: Can we say something that indicates that it's easy to build these and not just integrate? So maybe "...build or integrate result rerankers...?" Also, I think instead of saying "...as opposed to building plugins or creating additional logic elsewhere in your application..." we want to say that building processors could reduce complexity in building plugins used in search requests by allowing the functionality of a plugin to be composed by multiple processors. Might be too wordy, but search processors are not there to replace plugins.

File: _search-plugins/search-pipelines/rename-field-processor.md [hunk 22]
```diff
@@ -0,0 +1,214 @@
...
+parent: Search pipelines
+grand_parent: Search
+---
+
+# Rename field processor
+
+Rename field processor is part of search pipeline functionality. Search pipelines is an experimental feature. For updates on the progress of search pipelines, or if you want to leave feedback that could help improve the feature, see the associated [GitHub issue](https://github.com/opensearch-project/OpenSearch/issues/6278).    
+{: .warning}
+
+The `rename_field` search response processor intercepts a search response and renames the specified field. This is useful when your index and your application have different names for the same field. For example, if you rename a field in your index, the `rename_field` processor can change the new name to the old one before sending the response to your application.
```
Comment (natebower) [inline _search-plugins/search-pipelines/rename-field-processor.md:15]:
```suggestion
The `rename_field` search response processor intercepts a search response and renames the specified field. This is useful when your index and your application use different names for the same field. For example, if you rename a field in your index, the `rename_field` processor can change the new name to the old one before sending the response to your application.
```

File: _search-plugins/search-pipelines/rename-field-processor.md [hunk 23]
```diff
@@ -0,0 +1,214 @@
...
+## Request fields
+
+The following table lists all available request fields.
+
+Field | Data type | Description
+:--- | :--- | :---
+`field` | String | The field to rename. Required.
+`target_field` | String | The new field name. Required.
+`tag` | String | The processor's identifier. 
+`description` | String | Description of the processor.
```
Comment (natebower) [inline _search-plugins/search-pipelines/rename-field-processor.md:26]:
```suggestion
`description` | String | A description of the processor. 
```

File: _search-plugins/search-pipelines/rename-field-processor.md [hunk 24]
```diff
@@ -0,0 +1,214 @@
...
+`tag` | String | The processor's identifier. 
+`description` | String | Description of the processor. 
+
+## Example 
+
+The following example demonstrates using a search pipeline with a `rename_field` processor.
+
+### Setup
+
+Create an index `my_index` and index a document with the field `message`:
```
Comment (natebower) [inline _search-plugins/search-pipelines/rename-field-processor.md:34]:
```suggestion
Create an index named `my_index` and index a document with the field `message`:
```

File: images/search-pipelines.png
Comment (macohen) [inline images/search-pipelines.png:1]: great diagram!

File: _search-plugins/search-pipelines/rename-field-processor.md [hunk 25]
```diff
@@ -0,0 +1,214 @@
...
+      }
+    }
+  ]
+}
+```
+{% include copy-curl.html %}
+
+### Using a search pipeline
+
+Search for documents in `my_index` without search pipelines:
```
Comment (natebower) [inline _search-plugins/search-pipelines/rename-field-processor.md:66]: Same comment as in previous file

**Review by macohen — COMMENTED** [review]
I think there are a few changes in the way we describe pipelines/processors needed, but this looks great overall.

File: _search-plugins/search-pipelines/rename-field-processor.md [hunk 26]
```diff
@@ -0,0 +1,214 @@
...
+          "notification" : "This is a public message"
+        }
+      }
+    ]
+  }
+}
+```
+</details>
+
+You can also use the `fields` option to search for specific fields in the document:
```
Comment (natebower) [inline _search-plugins/search-pipelines/rename-field-processor.md:158]: Confirm that it shouldn't be "a" document.

File: _search-plugins/search-pipelines/script-processor.md [hunk 27]
```diff
@@ -0,0 +1,60 @@
...
+title: Script processor
+nav_order: 30
+has_children: false
+parent: Search pipelines
+grand_parent: Search
+---
+
+# Script processor
+
+Script processor is part of search pipeline functionality. Search pipelines is an experimental feature. For updates on the progress of search pipelines, or if you want to leave feedback that could help improve the feature, see the associated [GitHub issue](https://github.com/opensearch-project/OpenSearch/issues/6278).
```
Comment (natebower) [inline _search-plugins/search-pipelines/script-processor.md:12]:
```suggestion
The script processor is part of the functionality of a search pipeline, which is an experimental feature. For updates on the progress of the feature or if you want to leave feedback, see the associated [GitHub issue](https://github.com/opensearch-project/OpenSearch/issues/6278).    
```

File: _search-plugins/search-pipelines/script-processor.md [hunk 28]
```diff
@@ -0,0 +1,60 @@
...
+## Request fields
+
+The following table lists all available request fields.
+
+Field | Data type | Description
+:--- | :--- | :---
+`source` | Inline script | The script to run. Required.
+`lang` | String | The script language. Optional. Only `painless` is supported.
+`tag` | String | The processor's identifier. Optional.
+`description` | String | Description of the processor. Optional.
```
Comment (natebower) [inline _search-plugins/search-pipelines/script-processor.md:39]:
```suggestion
`description` | String | A description of the processor. Optional.
```

File: _search-plugins/search-pipelines/script-processor.md [hunk 29]
```diff
@@ -0,0 +1,60 @@
...
+
+## Example 
+
+The following request creates a search pipeline with a `script` request processor. The script limits score explanation to be included for one document only:
```
Comment (natebower) [inline _search-plugins/search-pipelines/script-processor.md:43]: The second sentence is unclear and needs revision. Do we mean something like "The script limits the explain operation to only one document"?

**Review by natebower — COMMENTED** [review]
@kolchfa-aws Please see my comments and changes and let me know if you have any questions. Thanks!

File: _search-plugins/search-pipelines/script-processor.md (hunk 27 above)
Comment (msfroh) [reply _search-plugins/search-pipelines/script-processor.md:12]: > The script processor is part of the functionality of a search pipeline

That's confusing.

A script processor is a type of processor that may be added to a search pipeline, if you want.

Comment (kolchfa-aws) [reply _search-plugins/search-pipelines/script-processor.md:12]: @msfroh I have put the same experimental warning on all files so they are uniform.

File: _search-plugins/search-pipelines/filter-query-processor.md (hunk 17 above)
Comment (kolchfa-aws) [reply _search-plugins/search-pipelines/filter-query-processor.md:24]: This is covered in the main page.

**Review by vagimeli — APPROVED** [review]
No changes. Looks good.

**Review by vagimeli — APPROVED** [review]
No changes. Looks good.
